txscript/pkscript.go

```diff
@@ -231,35 +238,24 @@ func computeNonWitnessPkScript(sigScript []byte) (PkScript, error) {
 
 // computeWitnessPkScript computes the script of an output by looking at the
 // spending input's witness.
+// IMPORTANT: With the addition of taproot, we can no longer say for certain
+// what kind of script the witness is in most cases. The only case in which we
+// can say for sure is when the witness data has an annex as the last push. In
+// that case, we can identify the script type, but we lack the ability to
+// reconstruct the script itself.
```

Comment:
i would say that the only case we know for certain is when the number of witness elements is exactly 1. In that case we know it is p2tr keypath (for now at least. not future proof). 

I dont know if we can say that that it is taproot just cause there is an annex starting with 0x50 cause surely there can be other p2w scripts (pre-taproot) that just happen to have a final witness element starting with 0x50?

Reply:
let me know what you think

Reply:
> I dont know if we can say that that it is taproot just cause there is an annex starting with 0x50 cause surely there can be other p2w scripts (pre-taproot) that just happen to have a final witness element starting with 0x50?

I guess that's true. Mkay.

> i would say that the only case we know for certain is when the number of witness elements is exactly 1. In that case we know it is p2tr keypath (for now at least. not future proof).

I'm unable to verify this. Is it impossible to construct a p2wsh script spent by a length 1 witness?

Reply:
> I guess that's true. Mkay.

Ok, I was wrong about this. I was not super familiar with how scripts are parsed. So yeah, if it starts with `0x50` then it is the annex

Reply:
Sorry to interject as a third party, but while I found https://github.com/lightninglabs/neutrino/pull/234#issuecomment-983185360 very enlightening, I think it would all fall apart when witness version 2 rolls around, like what happened with version 1 showed up and looked like p2wsh.
So while it sounds like much can be surmised from the witness stack, a hard fail based on the result of `computeWitnessPkScript` seems risky.

Reply:
> a hard fail based on the result of computeWitnessPkScript seems risky

yeah, especially when considering future witness versions like you said